File: packages/ai-engine/src/strategies/registry.ts
/**
 * StrategyRegistry — Plugin discovery and selection for reasoning strategies
 * Manages registered strategies and selects the best one for a given task context
 *
 * @author Subash Karki
 */
import type {
  ActivationScore,
  ReasoningStrategy,
  StrategyRegistryEntry,
  StrategyRole,
  TaskContext,
} from '../types/strategy.js';
import type { StrategyPerformanceStore } from './performance-store.js';

const FALLBACK_STRATEGY_ID = 'direct';
const MIN_ACTIVATION_THRESHOLD = 0.1;

export class StrategyRegistry {
  private entries = new Map<string, StrategyRegistryEntry>();
  private performanceStore: StrategyPerformanceStore | null = null;

  /** Attach a performance store for adaptive strategy scoring. */
  setPerformanceStore(store: StrategyPerformanceStore): void {
    this.performanceStore = store;
  }

  /** Register a strategy with an optional priority (higher = preferred on ties). */
  register(strategy: ReasoningStrategy, priority = 0): void {
    this.entries.set(strategy.id, { strategy, enabled: true, priority });
  }

  /** Remove a strategy entirely. */
  unregister(id: string): void {
    this.entries.delete(id);
  }

  /** Enable a previously disabled strategy. */
  enable(id: string): void {
    const entry = this.entries.get(id);
    if (entry) entry.enabled = true;
  }

  /** Disable a strategy without removing it. */
  disable(id: string): void {
    const entry = this.entries.get(id);
    if (entry) entry.enabled = false;
  }

  /**
   * Select the best strategy for a given task context.
   *
   * 1. Filter to enabled strategies
   * 2. Call shouldActivate(context) on each
   * 3. Sort by activation score DESC, then priority DESC
   * 4. Return the highest scorer
   * 5. If none score above the threshold, return the "direct" strategy as fallback
   */
  select(context: TaskContext): ReasoningStrategy {
    const scored = this.scoreAll(context);

    // Find highest scorer above threshold
    const best = scored.find((s) => s.score.score > MIN_ACTIVATION_THRESHOLD);
    if (best) return best.strategy;

    // Fallback to direct strategy
    const direct = this.entries.get(FALLBACK_STRATEGY_ID);
    if (direct) return direct.strategy;

    // If direct isn't registered either, return the first available
    const first = scored[0];
    if (first) return first.strategy;

    throw new Error('StrategyRegistry: no strategies registered');
  }

  /**
   * Return all strategies with their activation scores, sorted by
   * score DESC then priority DESC. Useful for observability / debugging.
   */
  selectAll(
    context: TaskContext,
  ): Array<{ strategy: ReasoningStrategy; score: ActivationScore }> {
    return this.scoreAll(context);
  }

  /** List every registered strategy entry. */
  getAll(): StrategyRegistryEntry[] {
    return [...this.entries.values()];
  }

  /** Lookup a single strategy entry by id. */
  get(id: string): StrategyRegistryEntry | undefined {
    return this.entries.get(id);
  }

  /**
   * Find the first enabled strategy entry that carries the given role tag.
   * Used by the Orchestrator to locate refiner/escalator strategies without
   * coupling to hard-coded strategy IDs.
   */
  getByRole(role: StrategyRole): StrategyRegistryEntry | undefined {
    for (const entry of this.entries.values()) {
      if (entry.enabled && entry.strategy.role === role) return entry;
    }
    return undefined;
  }

  // ---------------------------------------------------------------------------
  // Internal helpers
  // ---------------------------------------------------------------------------

  private scoreAll(
    context: TaskContext,
  ): Array<{
    strategy: ReasoningStrategy;
    score: ActivationScore;
    priority: number;
  }> {
    const results: Array<{
      strategy: ReasoningStrategy;
      score: ActivationScore;
      priority: number;
    }> = [];

    for (const entry of this.entries.values()) {
      if (!entry.enabled) continue;
      const rawScore = entry.strategy.shouldActivate(context);

      // Apply historical performance weight if available
      let adjustedScore = rawScore.score;
      if (this.performanceStore) {
        const weight = this.performanceStore.getHistoricalWeight(
          entry.strategy.id,
          context.complexity,
        );
        adjustedScore = rawScore.score * weight;
      }

      const score: ActivationScore = {
        score: adjustedScore,
        reason: rawScore.reason,
      };
      results.push({ strategy: entry.strategy, score, priority: entry.priority });
    }

    // Sort: highest score first, then highest priority first
    results.sort((a, b) => {
      const scoreDiff = b.score.score - a.score.score;
      if (scoreDiff !== 0) return scoreDiff;
      return b.priority - a.priority;
    });

    return results;
  }
}

File: v2/internal/ai/strategies/registry.go
// Package strategies provides AI prompt enhancement strategies.
// Registry manages strategy registration, activation scoring, and selection
// with optional historical performance weighting.
//
// Author: Subash Karki
package strategies

import (
	"sort"
	"sync"
)

// Strategy is implemented by each execution strategy. The registry calls
// ShouldActivate to score candidates and Enrich to apply the chosen strategy.
type Strategy interface {
	ID() string
	Name() string
	Description() string
	ShouldActivate(assessment TaskAssessment) (score float64, reason string)
	Enrich(message string, assessment TaskAssessment, graphCtx string) string
}

// StrategyInfo is a read-only snapshot of a registered strategy used by
// metadata-introspection callers (e.g. phantom_orchestrator_strategies).
type StrategyInfo struct {
	ID          string
	Name        string
	Description string
	Enabled     bool
	Priority    int
}

// registryEntry pairs a strategy with its enabled flag and static priority.
type registryEntry struct {
	strategy Strategy
	enabled  bool
	priority int
}

// Registry holds registered strategies and selects the best one for a task.
type Registry struct {
	mu         sync.RWMutex
	strategies []registryEntry
	perfStore  *PerformanceStore
}

// NewRegistry creates an empty strategy registry.
func NewRegistry() *Registry { return &Registry{} }

// Register adds a strategy with the given priority (higher = preferred on tie).
func (r *Registry) Register(s Strategy, priority int) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.strategies = append(r.strategies, registryEntry{strategy: s, enabled: true, priority: priority})
}

// SetPerformanceStore attaches a PerformanceStore for historical weighting.
func (r *Registry) SetPerformanceStore(ps *PerformanceStore) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.perfStore = ps
}

// GetAll returns metadata for every registered strategy in registration order.
// Used by introspection callers that need to enumerate available strategies
// without invoking selection.
func (r *Registry) GetAll() []StrategyInfo {
	r.mu.RLock()
	defer r.mu.RUnlock()
	out := make([]StrategyInfo, 0, len(r.strategies))
	for _, e := range r.strategies {
		out = append(out, StrategyInfo{
			ID:          e.strategy.ID(),
			Name:        e.strategy.Name(),
			Description: e.strategy.Description(),
			Enabled:     e.enabled,
			Priority:    e.priority,
		})
	}
	return out
}

// scored is an internal type for sorting candidates during selection.
type scored struct {
	strategy Strategy
	score    float64
	priority int
}

// Select returns the best strategy for the given assessment, or nil if none qualify.
func (r *Registry) Select(assessment TaskAssessment) Strategy {
	r.mu.RLock()
	defer r.mu.RUnlock()

	var candidates []scored
	for _, e := range r.strategies {
		if !e.enabled {
			continue
		}
		score, _ := e.strategy.ShouldActivate(assessment)

		// Apply historical performance weight when available.
		if r.perfStore != nil {
			weight := r.perfStore.GetHistoricalWeight(e.strategy.ID(), assessment.Complexity)
			score *= weight
		}

		if score > 0.1 {
			candidates = append(candidates, scored{e.strategy, score, e.priority})
		}
	}

	if len(candidates) == 0 {
		// Fallback to first registered strategy (should be Direct).
		if len(r.strategies) > 0 {
			return r.strategies[0].strategy
		}
		return nil
	}

	sort.Slice(candidates, func(i, j int) bool {
		if candidates[i].score != candidates[j].score {
			return candidates[i].score > candidates[j].score
		}
		return candidates[i].priority > candidates[j].priority
	})

	return candidates[0].strategy
}
